Don't try to use Orca for queries that it doesn't support

andr-sokolov opened the pull request:
Don't try to use Orca for queries that it doesn't support

On translating to DXL we can find out that the query is unsupported. The better
time to make this decision is analyse stage, because in this case we can run
Postgres planner without trying to run Orca and, as a consequence, significantly
reduce planning time.
Orca doesn't support gp_dist_random, queries on master-only tables, and function
exec locations except for any. It's easy to detect on analyse stage, that
a query contains these unsupported query properties.
The number of unsupported query properties to check before translating to DXL
will be increased later.

------------------------------
Before the patch
```
postgres=# select oid from pg_class where relname='pg_class';
 oid  
------
 1259
(1 row)

Time: 9,869 ms
postgres=# 
```

After the patch
```
postgres=# select oid from pg_class where relname='pg_class';
 oid  
------
 1259
(1 row)

Time: 1,995 ms
postgres=#
```

Don't plan simple inserts using Orca

andr-sokolov committed:
Don't try to use Orca for queries that it doesn't support

On translating to DXL we can find out that the query is unsupported. The better
time to make this decision is analyse stage, because in this case we can run
Postgres planner without trying to run Orca and, as a consequence, significantly
reduce planning time.
Orca doesn't support gp_dist_random, queries on master-only tables, and function
exec locations except for any. It's easy to detect on analyse stage, that
a query contains these unsupported query properties.
The number of unsupported query properties to check before translating to DXL
will be increased later.

andr-sokolov committed:
Don't try to use Orca for queries that it doesn't support

On translating to DXL we can find out that the query is unsupported. The better
time to make this decision is analyse stage, because in this case we can run
Postgres planner without trying to run Orca and, as a consequence, significantly
reduce planning time.
Orca doesn't support gp_dist_random, queries on master-only tables, and function
exec locations except for any. It's easy to detect on analyse stage, that
a query contains these unsupported query properties.
The number of unsupported query properties to check before translating to DXL
will be increased later.

leborchuk commented:
+1, Also we could use PG optimizer for insert values queries

Alena0704 committed:
Don't try to use Orca for lateral queries too

Alena0704 commented:
I think this patch looks promising. I rebased the patch and also propose to switch off Orca to prepare query plan (last commit).

Alena0704 committed:
Don't try to use Orca for lateral queries too

Orca doesn't support LATERAL, so a query using it is handed to Orca
only to fail during DXL translation and fall back to the Postgres
planner. As with the other unsupported properties detected at parse
analysis in the previous commit, recognize LATERAL there as well and
set usePostgresPlanner, so the Postgres planner runs directly without
the wasted Orca attempt.

Alena0704 committed:
Don't treat views as master-only tables when choosing the planner

The master-only detection fires on a NULL distribution policy, which
views have too, so every query over a view was needlessly pushed onto
the Postgres planner. That path keeps the view's range table entry for
permission checks, which made pgaudit emit an extra OBJECT record for
the view.

A view is expanded into its base relations before planning, so Orca
never scans the view itself. Exclude views (RELKIND_VIEW) from the check.

Alena0704 committed:
Copy usePostgresPlanner when copying a Query

usePostgresPlanner was added to Query but never registered in the node
support functions, so copyObject() on a Query silently dropped it.
EXPLAIN copies the query before planning, so an EXPLAIN of a query that
set the lost it and was still handed to Orca - which then failed and
fell back, defeating the optimization.

Alena0704 committed:
Update expected output for the master-only Orca fallback
